pgwidget/pgwidget_core.py

```diff
@@ -1049,8 +1114,76 @@ def draw_children(self):
         
        
 
-        
-    
+    def resize_table_by_rows_and_cols(self, rows, cols):
+        self.cols = cols
+        self._init_col_width_dict()
+        super().resize_table_by_rows_and_cols(rows, cols)
+
+        self.is_ready_for_updating_df_subset = True
+
+    def move_selected(self,direction):
+
+        if self.df is None:
+            super().move_selected(direction)
+            return
+
+        if self.selected_cell_index is not None:
+            i,j=self.table_cells[self.selected_cell_index].coor
+
+
+            if direction==1: #right
+                if j==self.cols-1 and i==self.rows:
+                    target_coordinates=(0,0)
+                elif j==self.cols-1:
+                    target_coordinates=(i+1, 0)
+                else:
+                    target_coordinates=(i,j+1)
+
+            if direction==2: #up
+                if i==1:
+                    if (self.camera_row_offset) > 0: #last cell in whole grid -> go to first cell
+                        target_coordinates=(i,j)
+                        self.scrollbar.shift_scrollbar(-self.move_scrollbar_by_one_cell())
+                    else:
+                        new_j = (j - 1) % self.cols
+                        target_coordinates = (self.rows, new_j)
+                        self.scrollbar.move_down()
+                    self.scrollbar.calculate_handle_ratio_position()
+                    self.is_ready_for_updating_df_subset = True
+
+                else:
+                    target_coordinates=(i-1, j)
+
+            if direction==3:
+                target_coordinates=(i,j-1) #left
+
+            if direction==4:#down
+                if i==self.rows:
+                    if (self.rows + self.camera_row_offset) < self.df.shape[0]: #last cell in whole grid -> go to first cell
+                        target_coordinates=(i,j)
+                        self.scrollbar.shift_scrollbar(self.move_scrollbar_by_one_cell())
+                    else:
+                        new_j = (j + 1) % self.cols
+                        target_coordinates = (1, new_j)
+                        self.scrollbar.move_up()
+                    self.scrollbar.calculate_handle_ratio_position()
+                    self.is_ready_for_updating_df_subset = True
+
+                else:
+                    target_coordinates=(i+1, j)
+
+            target_cell_index=self.find_cell_index(*target_coordinates)
+            if target_cell_index is not None:
+                self.deselect_all_cells()
+                self.select_cell(target_cell_index)
+            else:
+                print("Move cam, branch")
+                self.move_camera(*target_coordinates)
+
```

Comment:
This function needs to be refactored

Reply:
Inside of every "direction" branch should be only several quite generic functions - like "get_next_row_coordinates(...)", "shift_row(...)" -> Try harder please, move_selected can be implemented on max 20 rows of code - focus on remove nested if and else conditions. It is not easy to read right now.